src/python_testing/TC_DGGEN_2_1.py

```diff
@@ -0,0 +1,283 @@
+# === END CI TEST ARGUMENTS ===
+
+
+import asyncio
+import logging
+import time
+from typing import List
+
+import matter.clusters as Clusters
+from matter.testing.matter_testing import MatterBaseTest, TestStep, asserts, async_test_body, default_matter_test_main
+
+
+class TC_DGGEN_2_1_Py(MatterBaseTest):
+    """
+    [TC-DGGEN-2.1] Attributes (DUT as Server) — Python version with adjustments:
+      - No factory reset at start
+      - UpTime check and reboot before RebootCount++ check
+      - NetworkInterfaces validated in code
+      - TotalOperationalHours: verify >2 (manual precondition) and post-factory reset <=1
+    """
+
+    def desc_TC_DGGEN_2_1_Py(self) -> str:
+        return "[TC-DGGEN-2.1] Attributes (Server) — Pythonized + tweaks - REQUIRES device running >2h without factory reset"
+
+    def pics_TC_DGGEN_2_1_Py(self):
+        return ["DGGEN.S"]
+
+    def steps_TC_DGGEN_2_1_Py(self) -> List[TestStep]:
+        return [
+            TestStep(0,  "Manual pre-check: Device in use >2h without factory reset (tester confirms)"),
+            TestStep(1,  "Read TotalOperationalHours and verify >2"),
+            TestStep(2,  "Step 2a mod: Read and save RebootCount (boot_count1)"),
+            TestStep(3,  "Step 4a/4b: Read UpTime (uptime1), wait 10s, read UpTime (uptime2) and verify uptime2 >= uptime1"),
+            TestStep(4,  "Step 4c: Reboot DUT (manual/automated according to your environment), wait for reconnection"),
+            TestStep(5,  "Step 2b mod: Read RebootCount and verify boot_count == boot_count1 + 1"),
+            TestStep(6,  "UpTime post-reboot (uptime3) < uptime2"),
+            TestStep(7,  "Step 3 mod: Programmatically validate NetworkInterfaces"),
+            TestStep(8,  "Factory reset (manual), re-commission, verify TotalOperationalHours <= 1", is_commissioning=True),
+        ]
+
+    # ------- helpers -------
+
+    async def _read_attr(self, dev_ctrl, cluster_attr, endpoint=None, node_id=None):
+        return await self.read_single_attribute(
+            dev_ctrl=dev_ctrl,
+            node_id=node_id if node_id is not None else self.req_node_id,
+            endpoint=endpoint if endpoint is not None else self.endpoint,
+            attribute=cluster_attr
+        )
+
+    async def _read_reboot_count(self, dev_ctrl) -> int:
+        return int(await self._read_attr(dev_ctrl, Clusters.GeneralDiagnostics.Attributes.RebootCount))
+
+    async def _read_uptime(self, dev_ctrl) -> int:
+        return int(await self._read_attr(dev_ctrl, Clusters.GeneralDiagnostics.Attributes.UpTime))
+
+    async def _read_total_hrs(self, dev_ctrl) -> int:
+        return int(await self._read_attr(dev_ctrl, Clusters.GeneralDiagnostics.Attributes.TotalOperationalHours))
+
+    async def _read_net_ifaces(self, dev_ctrl):
+        return await self._read_attr(dev_ctrl, Clusters.GeneralDiagnostics.Attributes.NetworkInterfaces)
+
+    async def _wait_for_commissionee(self, node_id=None, timeout_s=120):
```

Comment:
You don't need to do this in the python scripts - any read will force a connection.